csrc/device_lower/pass/allocation.cpp

```diff
@@ -88,6 +88,158 @@ Val* getStrideOfGlobalMemoryTensor(TensorView* tv, int64_t alloc_dim) {
+    if (tv->getMemoryType() == MemoryType::Shared) {
+      if (std::any_of(
+              tv->getAllocationDomain().begin(),
+              tv->getAllocationDomain().end(),
+              [](IterDomain* allocation_domain) {
+                return dynamic_cast<Swizzle*>(
+                           allocation_domain->definition()) != nullptr ||
+                    allocation_domain->getParallelType() == ParallelType::Bulk;
+              }) ||
+          std::is_permutation(
+              tv->getLoopDomain().begin(),
+              tv->getLoopDomain().end(),
+              tv->getAllocationDomain().begin(),
+              tv->getAllocationDomain().end())) {
+        return true;
+      }
+
+      // Honor the set allocation domain if the tensor is used by a
+      // TMA store or MmaOp
+      if (std::ranges::any_of(tv->uses(), [](Expr* expr) {
+            return ir_utils::isCpAsyncBulkStore(expr) || expr->isA<MmaOp>();
+          })) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Helper function to check if an allocation domain should be excluded
+  // Returns the excluded ID if found, nullptr otherwise.
+  // An allocation domain should be excluded if it is not required to be
+  // allocated. For example:
+  // T2_s_float[iS6{2}, iS11{3}, iS12{4}, iB8{16}] ca_pos( 2 )
+  // logical domain : (iS6{2}, iS7{12}, iB8{16})
+  // allocation domain : (iS15{3}, iS16{4}, iS6{2}, iB8{16})
+  // contiguity: t t t t
+  //  Split: iS7{12} by factor 4 -> iS15{3}, iS16{4}
+  //  Split: iS7{12} by factor 4 -> iS11{3}, iS12{4}
+  // loop domain : (iS6{2}, iS11{3}, iS12{4}, iB8{16})
+  // Based on loop domain and compute pos, we don't need to allocate iS6{2}
+  // and iS11{3}. Then the corresponding allocation domains iS6{2} and
+  // iS15{3} should be excluded.
+  // iS6{2} is found directly by pointer comparison.
+  // iS15{3} is found by IdModel.
+  IterDomain* getExcludedAllocationDomain(
+      IterDomain* id,
+      const std::unordered_set<IterDomain*>& exclude_ca_ids) {
+    auto exclude_it = exclude_ca_ids.find(id);
+    if (exclude_it != exclude_ca_ids.end()) {
+      return *exclude_it;
+    }
+    return nullptr;
+  }
+
+  // Helper function to collect loop domains that should be excluded
+  std::unordered_set<IterDomain*> collectExcludedLoopDomains(
+      TensorView* tv,
+      int64_t allocation_pos) {
+    std::unordered_set<IterDomain*> exclude_ca_ids;
+    for (auto i : arange(allocation_pos)) {
+      auto ca_id = tv->axis(i);
+      if (!ir_utils::isMemorySharedAcross(
+              tv->getMemoryType(), ca_id->getParallelType())) {
+        exclude_ca_ids.insert(ca_id);
+      }
+    }
+    return exclude_ca_ids;
+  }
+
+  // select domains from allocation domain that should be allocated
+  std::pair<std::vector<IterDomain*>, std::vector<std::optional<bool>>>
+  usePresetAllocationDomain(
+      TensorView* tv,
+      const std::vector<ForLoop*>& for_loops) {
+    if (tv->getMemoryType() == MemoryType::Global) {
```

Comment:
Is my understanding correct that only changes made to the logic is inside `usePresetAllocationDomain`?

Reply:
> I can use a quick context / high level description on what's the motivation behind the refactor.

The previous function became too large and difficult to maintain. It includes logic to determine whether the preset allocation domain can be reused, how to reuse it if possible, and how to set a new allocation domain if not. The code contains many deeply nested if-else branches, making it hard to follow and maintain.